### ISSUE TITLE
Can't remove from forge huge repository

### ISSUE THREAD

**Issue by catap** [magit/forge#523]
An attempts to remove a huge repository from forge from https://github.com/macports/macports-ports fails with `Query timed out`.

Backtrace:
```elisp
Debugger entered--Lisp error: (emacsql-timeout "Query timed out" 30)
  signal(emacsql-timeout ("Query timed out" 30))
  #f(compiled-function (connection &optional timeout) "Block until CONNECTION is waiting for further input." #<bytecode -0x1332b5ae8745f3c1>)(#<forge-database forge-database-5a062c88>)
  apply(#f(compiled-function (connection &optional timeout) "Block until CONNECTION is waiting for further input." #<bytecode -0x1332b5ae8745f3c1>) #<forge-database forge-database-5a062c88> nil)
  emacsql-wait(#<forge-database forge-database-5a062c88>)
  #f(compiled-function (connection sql &rest args) #<bytecode -0x98d256621c96da4>)(#<forge-database forge-database-5a062c88> [:delete-from $i1 :where (= $i2 $s3)] repository id "Z2l0aHViLmNvbTowMTA6UmVwb3NpdG9yeTcwMzY1MTAw")
  apply(#f(compiled-function (connection sql &rest args) #<bytecode -0x98d256621c96da4>) (#<forge-database forge-database-5a062c88> [:delete-from $i1 :where (= $i2 $s3)] repository id "Z2l0aHViLmNvbTowMTA6UmVwb3NpdG9yeTcwMzY1MTAw"))
  (closure ((cl--args #<forge-database forge-database-5a062c88> [:delete-from $i1 :where (= $i2 $s3)] repository id "Z2l0aHViLmNvbTowMTA6UmVwb3NpdG9yeTcwMzY1MTAw") (cl--nmp . always) (cl--nm . #f(compiled-function (connection sql &rest args) #<bytecode -0x98d256621c96da4>)) t) (&rest args) (apply cl--nm (or args cl--args)))(#<forge-database forge-database-5a062c88> [:delete-from $i1 :where (= $i2 $s3)] repository id "Z2l0aHViLmNvbTowMTA6UmVwb3NpdG9yeTcwMzY1MTAw")
  apply((closure ((cl--args #<forge-database forge-database-5a062c88> [:delete-from $i1 :where (= $i2 $s3)] repository id "Z2l0aHViLmNvbTowMTA6UmVwb3NpdG9yeTcwMzY1MTAw") (cl--nmp . always) (cl--nm . #f(compiled-function (connection sql &rest args) #<bytecode -0x98d256621c96da4>)) t) (&rest args) (apply cl--nm (or args cl--args))) #<forge-database forge-database-5a062c88> [:delete-from $i1 :where (= $i2 $s3)] (repository id "Z2l0aHViLmNvbTowMTA6UmVwb3NpdG9yeTcwMzY1MTAw"))
  (mapcar #'closql--extern-unbound (apply cl--cnm connection sql (mapcar #'(lambda (arg) (if (stringp arg) (let (...) (set-text-properties 0 ... nil copy) copy) arg)) args)))
  (progn (mapcar #'closql--extern-unbound (apply cl--cnm connection sql (mapcar #'(lambda (arg) (if (stringp arg) (let ... ... copy) arg)) args))))
  (closure ((cl--cnm closure ((cl--args #<forge-database forge-database-5a062c88> [:delete-from $i1 :where (= $i2 $s3)] repository id "Z2l0aHViLmNvbTowMTA6UmVwb3NpdG9yeTcwMzY1MTAw") (cl--nmp . always) (cl--nm . #f(compiled-function (connection sql &rest args) #<bytecode -0x98d256621c96da4>)) t) (&rest args) (apply cl--nm (or args cl--args))) (cl--args #<forge-database forge-database-5a062c88> [:delete-from $i1 :where (= $i2 $s3)] repository id "Z2l0aHViLmNvbTowMTA6UmVwb3NpdG9yeTcwMzY1MTAw") (cl--nmp . always) (cl--nm . #f(compiled-function (connection sql &rest args) #<bytecode -0x98d256621c96da4>)) t) (cl--cnm connection sql &rest args) (progn (mapcar #'closql--extern-unbound (apply cl--cnm connection sql (mapcar #'(lambda ... ...) args)))))((closure ((cl--args #<forge-database forge-database-5a062c88> [:delete-from $i1 :where (= $i2 $s3)] repository id "Z2l0aHViLmNvbTowMTA6UmVwb3NpdG9yeTcwMzY1MTAw") (cl--nmp . always) (cl--nm . #f(compiled-function (connection sql &rest args) #<bytecode -0x98d256621c96da4>)) t) (&rest args) (apply cl--nm (or args cl--args))) #<forge-database forge-database-5a062c88> [:delete-from $i1 :where (= $i2 $s3)] repository id "Z2l0aHViLmNvbTowMTA6UmVwb3NpdG9yeTcwMzY1MTAw")
  apply((closure ((cl--cnm closure ((cl--args #<forge-database forge-database-5a062c88> [:delete-from $i1 :where (= $i2 $s3)] repository id "Z2l0aHViLmNvbTowMTA6UmVwb3NpdG9yeTcwMzY1MTAw") (cl--nmp . always) (cl--nm . #f(compiled-function (connection sql &rest args) #<bytecode -0x98d256621c96da4>)) t) (&rest args) (apply cl--nm (or args cl--args))) (cl--args #<forge-database forge-database-5a062c88> [:delete-from $i1 :where (= $i2 $s3)] repository id "Z2l0aHViLmNvbTowMTA6UmVwb3NpdG9yeTcwMzY1MTAw") (cl--nmp . always) (cl--nm . #f(compiled-function (connection sql &rest args) #<bytecode -0x98d256621c96da4>)) t) (cl--cnm connection sql &rest args) (progn (mapcar #'closql--extern-unbound (apply cl--cnm connection sql (mapcar #'(lambda ... ...) args))))) (closure ((cl--args #<forge-database forge-database-5a062c88> [:delete-from $i1 :where (= $i2 $s3)] repository id "Z2l0aHViLmNvbTowMTA6UmVwb3NpdG9yeTcwMzY1MTAw") (cl--nmp . always) (cl--nm . #f(compiled-function (connection sql &rest args) #<bytecode -0x98d256621c96da4>)) t) (&rest args) (apply cl--nm (or args cl--args))) (#<forge-database forge-database-5a062c88> [:delete-from $i1 :where (= $i2 $s3)] repository id "Z2l0aHViLmNvbTowMTA6UmVwb3NpdG9yeTcwMzY1MTAw"))
  (let ((cl--cnm #'(lambda (&rest args) (apply cl--nm (or args cl--args))))) (apply #'(lambda (cl--cnm connection sql &rest args) (progn (mapcar #'closql--extern-unbound (apply cl--cnm connection sql (mapcar ... args))))) cl--cnm cl--args))
  (closure ((cl--nmp . always) (cl--nm . #f(compiled-function (connection sql &rest args) #<bytecode -0x98d256621c96da4>)) t) (&rest cl--args) "\n\n(fn CONNECTION SQL &rest ARGS)" (let ((cl--cnm #'(lambda (&rest args) (apply cl--nm ...)))) (apply #'(lambda (cl--cnm connection sql &rest args) (progn (mapcar ... ...))) cl--cnm cl--args)))(#<forge-database forge-database-5a062c88> [:delete-from $i1 :where (= $i2 $s3)] repository id "Z2l0aHViLmNvbTowMTA6UmVwb3NpdG9yeTcwMzY1MTAw")
  apply((closure ((cl--nmp . always) (cl--nm . #f(compiled-function (connection sql &rest args) #<bytecode -0x98d256621c96da4>)) t) (&rest cl--args) "\n\n(fn CONNECTION SQL &rest ARGS)" (let ((cl--cnm #'(lambda (&rest args) (apply cl--nm ...)))) (apply #'(lambda (cl--cnm connection sql &rest args) (progn (mapcar ... ...))) cl--cnm cl--args))) #<forge-database forge-database-5a062c88> ([:delete-from $i1 :where (= $i2 $s3)] repository id "Z2l0aHViLmNvbTowMTA6UmVwb3NpdG9yeTcwMzY1MTAw"))
  emacsql(#<forge-database forge-database-5a062c88> [:delete-from $i1 :where (= $i2 $s3)] repository id "Z2l0aHViLmNvbTowMTA6UmVwb3NpdG9yeTcwMzY1MTAw")
  (let ((key (eieio-oref-default obj 'closql-primary-key))) (emacsql (closql--oref obj 'closql-database) [:delete-from $i1 :where (= $i2 $s3)] (eieio-oref-default obj 'closql-table) key (closql--oref obj key)))
  (progn (let ((key (eieio-oref-default obj 'closql-primary-key))) (emacsql (closql--oref obj 'closql-database) [:delete-from $i1 :where (= $i2 $s3)] (eieio-oref-default obj 'closql-table) key (closql--oref obj key))))
  (closure (t) (obj) (progn (let ((key (eieio-oref-default obj 'closql-primary-key))) (emacsql (closql--oref obj 'closql-database) [:delete-from $i1 :where (= $i2 $s3)] (eieio-oref-default obj 'closql-table) key (closql--oref obj key)))))(#<forge-github-repository forge-github-repository-5a087dca>)
  apply((closure (t) (obj) (progn (let ((key (eieio-oref-default obj 'closql-primary-key))) (emacsql (closql--oref obj 'closql-database) [:delete-from $i1 :where (= $i2 $s3)] (eieio-oref-default obj 'closql-table) key (closql--oref obj key))))) #<forge-github-repository forge-github-repository-5a087dca> nil)
  closql-delete(#<forge-github-repository forge-github-repository-5a087dca>)
  forge-remove-repository("github.com" "macports" "macports-ports")
  funcall-interactively(forge-remove-repository "github.com" "macports" "macports-ports")
  command-execute(forge-remove-repository record)
  counsel-M-x-action("forge-remove-repository")
  ivy-call()
  ivy-read("M-x " ("toggle-debug-on-error" "forge-remove-repository" "forge-list-repositories" "forge-visit-repository" "sort-lines" "diff-buffer-with-file" "find-grep" "replace-string" "makefile-mode" "c-mode" "treemacs" "code-review-start" "cd" "wl" "5x5" "amx" "arp" "dbx" "dig" "erc" "ert" "eww" "ftp" "gdb" "irc" "jdb" "lsp" "man" "mpc" "pdb" "pwd" "rsh" "sdb" "xdb" "calc" "diff" "ffap" "gnus" "grep" "help" "ielm" "info" "life" "lsp!" "mail" "mpuz" "ping" "pong" "talk" "term" ...) :predicate counsel--M-x-externs-predicate :require-match t :history counsel-M-x-history :action counsel-M-x-action :keymap (keymap (67108908 . counsel--info-lookup-symbol) (67108910 . counsel-find-symbol)) :initial-input nil :caller counsel-M-x)
  counsel-M-x()
  funcall-interactively(counsel-M-x)
  command-execute(counsel-M-x)
```

**Comment by tarsius** [comment]
You should be able to deal with that using `(setq emacsql-global-timeout 1000)`.

**Comment by catap** [comment]
Well, it works for about 20 minutes.

I feel that it is wrong way to remove it. But if you think otherwise feel free to close an issue.

**Comment by tarsius** [comment]
That's pretty extreme. I never had to increase the timeout myself.

I am working on new sqlite backends for emacsql at https://github.com/magit/emacsql/tree/sqlite-backends. Except for some outstanding improvement concerning error reporting, they are ready. I have been using them for months.

They are faster. It would be interesting to learn how much faster in such an extreme case.

**Comment by catap** [comment]
@tarsius switched to your new backend. 30 seconds isn't enough to delete such repository. Trying with larges timeout.

**Comment by catap** [comment]
I've measured by noise that my laptop makes and would like to say that it hadn't changed anything much. Maybe it faster, maybe not.

But it still more than 10-15 minutes.

**Comment by tarsius** [comment]
A repository is deleted using something like

```sql
DELETE FROM repository
WHERE id = '"Z2l0aHViLmNvbTowMTA6UmVwb3NpdG9yeTg2OTE5NzI="';
```

which seems reasonable. But deletes do cascade, meaning that lots rows have to be deleted from lots of tables, and that apparently takes a lot of time for huge repositories. Beside the size of the data that has to be deleted, the total size has an effect too.

I don't think there is much we can do about this.

Is the db located on a spinning hard disk?

**Comment by catap** [comment]
@tarsius nope, it is SSD. Maybe it misses indexes?

**Comment by catap** [comment]
BTW, when timeout is happened it haven't kill `emacsql-sqlite` process which is continue to works.

**Comment by tarsius** [comment]
Closing because I was never able to reproduce and don't think this can happen with the new back-ends.

> switched to your new backend.

I suspect you did not succeed in doing that, because...

> BTW, when timeout is happened it haven't kill `emacsql-sqlite` process which is continue to works.

... when using one of the new back-ends, there is no "`emacsql-sqlite` process", and ...

> 30 seconds isn't enough to delete such repository. Trying with larges timeout.

... the new back-ends do *not* abort a request if it longer than `emacsql-global-timeout`. 

(They aborted if access was blocked longer than that by another process using the same db. I have just added a separate variable to separately control this distinct timeout type.)

If there indeed is a bug as described when using the legacy `sqlite` back-end, then this is a `wontfix` now; the new back-ends are available in the v4 release and the legacy back-end will be removed soon.

`(type-of (oref (forge-db) connection))` can now be used to check what back-end is being used.
